kubespawner/objects.py

```diff
@@ -566,3 +610,103 @@ def make_ingress(
     )
 
     return endpoint, service, ingress
+
+
+def make_secret(
+    name,
+    env=None,
+    labels=None,
+    annotations=None,
+):
+    """
+    Make a k8s secret specification for mount ssl credentials for running a user notebook.
+
+    Parameters
+    ----------
+    name:
+        Name of the secret. Must be unique within the namespace the object is
+        going to be created in.
+    env:
+        Dictionary of environment variables.
+    labels:
+        Labels to add to the spawned pod.
+    annotations:
+        Annotations to add to the spawned pod.
+
+    """
+
+    if not ssl_enabled(env):
+        return None
+
+    secret = V1Secret()
+    secret.kind = "Secret"
+    secret.api_version = "v1"
+    secret.metadata = V1ObjectMeta()
+    secret.metadata.name = name
+    secret.metadata.annotations = (annotations or {}).copy()
+    secret.metadata.labels = (labels or {}).copy()
+
+    secret.data = {}
+    with open(env['JUPYTERHUB_SSL_KEYFILE'], 'r') as file:
```

Comment:
We should pass these all in manually, or by passing in an appropriate object. We shouldn't use env vars like this.

Reply:
@sstarcher: this is the current implementation in this section, which is running in JupyterHub / KubeSpawner.

```
    secret.data = {}
    user_ssl_key = "{}/user-{}/user-{}.key".format(ssl_directory, username, username)
    with open(user_ssl_key, 'r') as file:
        encoded = base64.b64encode(file.read().encode("utf-8"))
        secret.data['ssl.key'] = encoded.decode("utf-8")
```

This section should still not rely on a environment variable, as it is within the JupyterHub pod, and this section is where we will rely on information in `self.certs_path` instead or `local_spawner_certs_path` as we may have updated `self.certs_path` to represent the location for the user.

Reply:
I'll mark this as unresolved to let you press resolve if you agree @sstarcher!